clippy_lints/src/swap.rs

```diff
@@ -220,6 +221,92 @@ fn check_suspicious_swap(cx: &LateContext<'_>, block: &Block<'_>) {
                     });
             }
         }
+
+        if let StmtKind::Local(first) = first_.kind && let StmtKind::Local(second) = second_.kind {
+            parse_mutual(cx, first, second);
+        }
+
+        if let StmtKind::Local(first) = first_.kind && let StmtKind::Semi(second) = second_.kind {
+            parse(cx, first, second);
+        }
```

Comment:
And then we can remove this, there should be only one place where we're emitting the lint

```suggestion
```

Reply:
@Alexendoo I have read the comment and am stucking of refactoring the [L:238~L:293](https://github.com/rust-lang/rust-clippy/pull/8945/files#diff-342cbd825f201119c85fd8892ed86fb369d63b914c1e1e0179a40ef10215bd44R238-R293).How can I simplify these functions...
```
enum ExprOrLocal<'a> {
    Expr(&'a Expr<'a>),
    Local(&'a Local<'a>),
}
// Checking both statement type but probably I'm doing wrong 
fn parse<'a, 'hir>(first: &'a Stmt<'hir>, second: &'a Stmt<'hir>) -> Option<(&'a Local<'hir>, &'a ExprOrLocal<'hir>)> { // 
    if let StmtKind::Local(first_hs) = first.kind {
        if let StmtKind::Local(second_hs) = second.kind {
            if first_hs.span.ctxt() == second_hs.span.ctxt() {
                Some((first_hs, second_hs));
            }
        }
    } else if let StmtKind::Local(first_hs) = first.kind {
        if let StmtKind::Semi(second_hs) = second.kind {
            if first_hs.span.ctxt() == second_hs.span.ctxt() {
                Some((first_hs, second_hs));
            }
        }
    }
    None
}

if let Some((lhs1, rhs1)) = parse(first, second) {
  // Trying to make L:238 ~ L:293 DRY in here.
  // ref: https://github.com/rust-lang/rust-clippy/pull/8945/files#diff-342cbd825f201119c85fd8892ed86fb369d63b914c1e1e0179a40ef10215bd44R238-R293) DRY in here.
}
```

Reply:
`parse` should only take a single `Stmt` which you're breaking down into either `let {ident} = {expr}` or `{expr} = {expr}`

`ExprOrLocal::Local` should be an `Ident`, not `Local`, you could rename it to `ExprOrIdent` as well to make that clearer

You'd call it twice, once with the first statement and once with the second. Then you'd write a function

```rust
fn is_same(cx: .., lhs: ExprOrIdent, rhs: &Expr) -> bool { .. }
```

to call instead of `eq_expr_value` now that `lhs0` and `lhs1` are `ExprOrIdent`s instead of `Expr`s

Reply:
@Alexendoo Thank you so much for your advice. I have updated the code and trying to fix these changes now.
https://github.com/rust-lang/rust-clippy/blob/b2482d2db561b456661a910e46c10db45b0abcc5/tests/ui/crate_level_checks/no_std_swap.stderr#L6
https://github.com/rust-lang/rust-clippy/blob/b2482d2db561b456661a910e46c10db45b0abcc5/tests/ui/swap.fixed#L117